### PR TITLE
Bump Nessie to 0.8.2 + related changes

### PR DESCRIPTION
**Pull request by snazy**
* Bump Nessie to 0.8.2 + replace Gradle plugin with new JUnit extension
  * Replace Gradle plugin with new JUnit extension.
    See [Add JAX-RS tests and add JUnit/Jupyter extension](https://github.com/projectnessie/nessie/pull/1566)

* Apply changes to Iceberg required by API changes in Nessie:
  * [Re-introduce wrapper classes for query params of CommitLog/Entries](https://github.com/projectnessie/nessie/pull/1595)
  * [Server-side commit range filtering](https://github.com/projectnessie/nessie/pull/1596)
  * [Add hashOnRef query param to support time travel on a named ref](https://github.com/projectnessie/nessie/pull/1589)
  * [Only accept NamedRefs in REST API](https://github.com/projectnessie/nessie/pull/1583)

* Bugfix: must send the Contents.id of the existing table
    Nessie's `Contents.id` is a random ID generated when the `Contents.Key` is first used (think:
    CREATE TABLE) and must not be changed. This change addresses a bug in the Iceberg-Nesie code
    that caused a new id for every change.

    Throw `CommitStateUnknownException` for `renameTable` as well
    
    Follow-up of #2515

* Fix race-condition & save one roundtrip to Nessie during "commit"
    
    When commiting a change, the Nessie-API now returns the hash of the commit for the change.
    This returned hash should then be used as the "expected hash" for the next commit.
    
    The previous approach was to commit the change to Nessie and then do another request to
    retrieve the new hash of HEAD.
    
    This old approach is prone to a race condition, namely when another commit happens after
    "this" commit but before retrieving the "new HEAD", so "this" instance would wrongly
    ignore the other commit's changes during conflict checks.
    
    See [Let VersionStore.create()+commit() return the current hash](https://github.com/projectnessie/nessie/pull/1089)

### CODE REVIEW

**Review by nastra — APPROVED** [review]
changes LGTM

**Comment by rymurr** [comment]
Looks like this needs a rebase. @snazy do you still want to update to `0.6.1`?

**Comment by snazy** [comment]
Yup, updated this PR to bump Nessie to 0.6.1.

File: versions.props [hunk 1]
```diff
@@ -19,7 +19,7 @@ org.apache.arrow:arrow-memory-netty = 2.0.0
 com.github.stephenc.findbugs:findbugs-annotations = 1.3.9-1
 software.amazon.awssdk:* = 2.15.7
 org.scala-lang:scala-library = 2.12.10
-org.projectnessie:* = 0.5.1
+org.projectnessie:* = 0.6.1
 javax.ws.rs:javax.ws.rs-api = 2.1.1
 io.quarkus:* = 1.13.1.Final
```
Comment (rymurr) [inline versions.props:24]: does `0.6.1` have a new version of quarkus?

Comment (snazy) [reply versions.props]: Ah! Good catch! Updated it.

**Review by nastra — APPROVED** [review]

**Comment by rymurr** [comment]
@snazy looks like the build failed for java8. Presumably its the recent chagne to nessie server to not support java8?

**Comment by rymurr** [comment]
I see no rush to get this patch in. The only deadline is the 0.12.0 release. So lets fix the tests correctly.

Lets go with the uber-jar too. I don't want to complicate the CI and introduce another build dependency.

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 2]
```diff
@@ -350,8 +369,9 @@ private UpdateableReference loadReference(String requestedRef) {
 
   private Stream<TableIdentifier> tableStream(Namespace namespace) {
     try {
+      EntriesParams entriesParam = EntriesParams.empty();
```
Comment (nastra) [inline nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:372]: nit: let's inline this one

File: versions.props [hunk 3]
```diff
@@ -18,13 +18,11 @@ org.apache.arrow:arrow-memory-netty = 2.0.0
 com.github.stephenc.findbugs:findbugs-annotations = 1.3.9-1
 software.amazon.awssdk:* = 2.15.7
 org.scala-lang:scala-library = 2.12.10
-org.projectnessie:* = 0.5.1
-javax.ws.rs:javax.ws.rs-api = 2.1.1
-io.quarkus:* = 1.13.1.Final
+org.projectnessie:* = 0.7.1-SNAPSHOT
```
Comment (nastra) [inline versions.props:21]: assuming this will be updated with the proper release version

File: nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java [hunk 4]
```diff
@@ -44,6 +45,15 @@ public boolean refresh() throws NessieNotFoundException {
     return !oldReference.equals(reference);
   }
 
+  public void updateReference(Reference ref) {
+    Objects.requireNonNull(ref);
+    if (this.reference != null && this.reference.getClass() != ref.getClass()) {
```
Comment (nastra) [inline nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java:50]: can `this.reference` ever be `null`?

**Review by nastra — APPROVED** [review]
LGTM, with one question

File: versions.props (hunk 3 above)
Comment (snazy) [reply versions.props:21]: yep - that's a temporary commit in this PR

File: nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java (hunk 4 above)
Comment (snazy) [reply nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java]: Probably not.
Added some `@Nonnull` and removed that check.

**Comment by snazy** [comment]
Tests pass locally w/ Java 11 + Java 8 against latest in-dev Nessie.

File: nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java [hunk 5]
```diff
@@ -44,6 +48,15 @@ public boolean refresh() throws NessieNotFoundException {
     return !oldReference.equals(reference);
   }
 
+  public void updateReference(Reference ref) {
+    Objects.requireNonNull(ref);
+    if (this.reference.getClass() != ref.getClass()) {
```
Comment (nastra) [inline nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java:53]: nit: you could use `Preconditions.checkArgument(...)`

**Review by nastra — APPROVED** [review]

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 6]
```diff
@@ -174,26 +177,31 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
...
     try {
       Tasks.foreach(contents)
           .retry(5)
           .stopRetryOn(NessieNotFoundException.class)
           .throwFailureWhenFinished()
           .onFailure((c, exception) -> refresh())
           .run(c -> {
-            client.getTreeApi().commitMultipleOperations(reference.getAsBranch().getName(), reference.getHash(), c);
-            refresh(); // note: updated to reference.updateReference() with Nessie 0.6
+            Branch branch = client.getTreeApi().commitMultipleOperations(reference.getAsBranch().getName(),
```
Comment (rymurr) [inline nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:188]: This in my honest opinion should be a different change. This is atomic imho

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 7]
```diff
@@ -225,8 +233,13 @@ public void renameTable(TableIdentifier from, TableIdentifier toOriginal) {
           .throwFailureWhenFinished()
           .onFailure((c, exception) -> refresh())
           .run(c -> {
-            client.getTreeApi().commitMultipleOperations(reference.getAsBranch().getName(), reference.getHash(), c);
-            refresh();
+            Branch branch = client.getTreeApi().commitMultipleOperations(reference.getAsBranch().getName(),
```
Comment (rymurr) [inline nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:236]: imho this is a different change too but im less concerned about this.

Why do we check if the branch is null...when can it be null

File: nessie/src/main/java/org/apache/iceberg/nessie/TableReference.java [hunk 8]
```diff
@@ -21,25 +21,28 @@
...
 import java.time.Instant;
 import java.util.List;
+import javax.annotation.Nonnull;
 import org.apache.iceberg.catalog.TableIdentifier;
 import org.apache.iceberg.relocated.com.google.common.base.Splitter;
 
 public class TableReference {
 
   private static final Splitter BRANCH_NAME_SPLITTER = Splitter.on("@");
+  @Nonnull
```
Comment (rymurr) [inline nessie/src/main/java/org/apache/iceberg/nessie/TableReference.java:31]: what does this do?

**Review by rymurr — COMMENTED** [review]
I think there are two things:

1) more documentation of the individual changes - some aren't clear
2) separate into a few changes

File: nessie/src/main/java/org/apache/iceberg/nessie/TableReference.java (hunk 8 above)
Comment (snazy) [reply nessie/src/main/java/org/apache/iceberg/nessie/TableReference.java:31]: Been requested earlier during review to avoid unnecessary null-checkes

File: versions.props [hunk 9]
```diff
@@ -18,13 +18,11 @@ org.apache.arrow:arrow-memory-netty = 2.0.0
...
 org.scala-lang:scala-library = 2.12.10
-org.projectnessie:* = 0.5.1
-javax.ws.rs:javax.ws.rs-api = 2.1.1
-io.quarkus:* = 1.13.1.Final
+org.projectnessie:* = 0.8.2
 
 # test deps
-org.junit.vintage:junit-vintage-engine = 5.7.2
-org.junit.jupiter:junit-jupiter = 5.7.2
+org.junit.vintage:* = 5.7.2
```
Comment (rymurr) [inline versions.props:24]: why this change?

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 10]
```diff
@@ -225,8 +233,13 @@ public void renameTable(TableIdentifier from, TableIdentifier toOriginal) {
...
+                reference.getHash(), c);
+            if (branch != null) {
```
[the remark this answers is not in the document]
Comment (snazy) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:238]: Old Nessie versions could. 0.8.x never. I'll remove these checks.

File: nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java [hunk 11]
```diff
@@ -44,6 +45,15 @@ public boolean refresh() throws NessieNotFoundException {
...
+      throw new IllegalArgumentException("Expected " + this.reference.getClass().getName() +
```
[the remark this answers is not in the document]
Comment (snazy) [reply nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java:51]: yea - overly cautious

File: versions.props (hunk 9 above)
Comment (snazy) [reply versions.props:24]: the vintage one? not really necessary, but also doesn't harm. reverted.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java [hunk 12]
```diff
@@ -79,31 +79,34 @@ public void testUpdateCatalogs() {
...
     Assertions.assertThat(metadataLocation(catalog, TableIdentifier.of("test-ns", "table1@test")))
         .isEqualTo(metadataLocation(testCatalog, tableIdentifier1));
 
-    // asking for table@branch#hash gives expected regardless of catalog
-    Assertions.assertThat(metadataLocation(catalog, TableIdentifier.of("test-ns", "table1@" + mainHash)))
+    // Asking for table@branch gives expected regardless of catalog.
+    // Earlier versions used "table1@" + tree.getReferenceByName("main").getHash() before, but since
+    // Nessie 0.8.2 the branch name became mandatory and specifying a hash within a branch is not
+    // possible.
+    Assertions.assertThat(metadataLocation(catalog, TableIdentifier.of("test-ns", "table1@" + mainName)))
```
[the remark this answers is not in the document]
Comment (snazy) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java:109]: Opened #2830

**Review by rymurr — APPROVED** [review]
Thanks @snazy !!
